storm-core/src/jvm/org/apache/storm/messaging/local/Context.java

```diff
@@ -82,31 +86,73 @@ public void close() {
 
     private static class LocalClient implements IConnection {
         private final LocalServer _server;
+        //Messages sent before the server registered a callback
+        private final LinkedBlockingQueue<TaskMessage> _pendingDueToUnregisteredServer;
+        private final ScheduledExecutorService _pendingFlusher;
 
         public LocalClient(LocalServer server) {
             _server = server;
+            _pendingDueToUnregisteredServer = new LinkedBlockingQueue<>();
+            _pendingFlusher = Executors.newScheduledThreadPool(1, new ThreadFactory(){
+                @Override
+                public Thread newThread(Runnable runnable) {
+                    Thread thread = new Thread(runnable);
+                    thread.setName("LocalClientFlusher-" + thread.getId());
+                    thread.setDaemon(true);
+                    return thread;
+                }
+            });
+            _pendingFlusher.scheduleAtFixedRate(new Runnable(){
+                @Override
+                public void run(){
+                    try {
+                        //Ensure messages are flushed even if no more sends are performed
+                        flushPending();
+                    } catch (Throwable t) {
+                        LOG.error("Uncaught throwable in pending message flusher thread, messages may be lost", t);
+                        throw t;
```

Comment:
It does, for some reason. It seems like the compiler actually checks what's being thrown in the try clause before complaining about this throw.
This compiles:
```
Runnable r = new Runnable() {
            @Override
            public void run() {
                try {
                    throw new RuntimeException("Something broke");
                } catch (Throwable t) {
                    throw t;
                }
            }

        };
        r.run();
```
and this doesn't:
```
Runnable r = new Runnable() {
            @Override
            public void run() {
                try {
                    throw new Exception("Something broke");
                } catch (Throwable t) {
                    throw t;
                }
            }

        };
        r.run();
```
I'll wrap the throwable in a RuntimeException.